runner/src/main.rs

```diff
@@ -0,0 +1,199 @@
+//! A utility binary to run tests and orchestrate examples and other tools within the repository,
+//! for local development and CI.
+//!
+//! To invoke, run the following command from the root of the repository:
+//!
+//! ```
+//! cargo run --manifest-path=./runner/Cargo.toml
+//! ```
+
+use colored::*;
+use std::io::Write;
+use std::path::PathBuf;
+use std::process::Command;
+
+/// Return whether to ignore the specified path. This is used by the `walker` package to efficiently
+/// avoid descending into blacklisted directories.
+fn is_ignored_path(path: &PathBuf) -> bool {
+    path.starts_with("./bazel-cache")
+        || path.starts_with("./cargo-cache")
+        || path.starts_with("./third_party")
+        || path.starts_with("./.git")
+        || path.ends_with("target") // Rust artifacts.
+}
+
+fn is_ignored_entry(entry: &walkdir::DirEntry) -> bool {
+    let path = entry.clone().into_path();
+    is_ignored_path(&path)
+}
+
+/// Return an iterator of all the first-party and non-ignored files in the repository, which can be
+/// then additionally filtered by the caller.
+fn source_files() -> impl Iterator<Item = PathBuf> {
+    let walker = walkdir::WalkDir::new(".").into_iter();
+    walker
+        .filter_entry(|e| !is_ignored_entry(e))
+        .filter_map(Result::ok)
+        .map(|e| e.into_path())
+}
+
+/// Return whether the provided path refers to a Bazel file (`BUILD`, `WORKSPACE`, or `*.bzl`)
+fn is_bazel_file(path: &PathBuf) -> bool {
+    let filename = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
+    filename == "BUILD" || filename == "WORKSPACE" || filename.ends_with(".bzl")
+}
+
+/// Return whether the provided path refers to a markdown file (`*.md`)
+fn is_markdown_file(path: &PathBuf) -> bool {
+    let filename = path.file_name().and_then(|s| s.to_str()).unwrap_or("");
+    filename.ends_with(".md")
+}
+
+fn main() {
+    let step = Step::root();
+    {
+        let step = step.with_prefix("buildifier");
+        for entry in source_files().filter(is_bazel_file) {
```

Comment:
I guess we could accumulate `source_files()` into a `Vec` once and for all, but only if re-scanning seems to be slow.

Reply:
Yeah I thought about that but it does not seem worth it for now.